src/redux/books/books.js

```diff
@@ -1,37 +1,81 @@
-const ADD_BOOK = 'ADD_BOOK';
-const REMOVE_BOOK = 'REMOVE_BOOK';
-const books = [
-  { id: 1, author: 'Book-1 author', title: 'Book-1' },
-  { id: 2, author: 'Book-2 author', title: 'Book-2' },
-  { id: 3, author: 'Book-3 author', title: 'Book-3' },
-  { id: 4, author: 'Book-4 author ', title: 'Book-4' },
-];
-
-export const AddBookFunction = (NewBook) => ({
+const ADD_BOOK = 'bookStore/books/ADD_BOOK';
+const REMOVE_BOOK = 'bookStore/books/REMOVE_BOOK';
+const FETCH_BOOKS = 'bookstore/books/FETCH_BOOKS';
+const urlLink = 'https://us-central1-bookstore-api-e63c8.cloudfunctions.net/bookstoreApi/apps/IEXNcgPY8gfuqg4iq5ZV/books';
+
+const initialState = [];
+
+export const addBook = (payload) => ({
   type: ADD_BOOK,
-  payload: NewBook,
+  payload,
 });
-export const RemoveBookFunction = (id) => ({
+
+export const removeBook = (payload) => ({
   type: REMOVE_BOOK,
-  id,
+  payload,
 });
 
-const BooksReducer = (state = books, action) => {
+export const fetchBook = (payload) => ({
+  type: FETCH_BOOKS,
+  payload,
+});
+
+const BookReducer = (state = initialState, action) => {
   switch (action.type) {
     case ADD_BOOK:
-      return [
-        ...state,
-        {
-          id: Date.now(),
-          title: action.payload.title,
-          author: action.payload.author,
-        },
-      ];
+      return [...state, action.payload];
+
     case REMOVE_BOOK:
-      return state.filter((book) => book.id !== action.id);
+      return state.filter((book) => book.id !== action.payload);
+
+    case FETCH_BOOKS:
+      return action.payload;
+
     default:
       return state;
   }
 };
 
-export default BooksReducer;
+export const fetchBooksList = () => async (dispatch) => {
+  const booksList = await fetch(`${urlLink}`).then((response) => response.json());
+  const booksID = Object.keys(booksList);
+  const formBooks = [];
+  booksID.map((key) => formBooks.push({
+    id: key,
+    title: booksList[key][0].title,
+    author: booksList[key][0].author,
+    category: booksList[key][0].category,
+  }));
+  dispatch(fetchBook(formBooks));
+};
```

Comment:
- [ ] As it is in the project requirements, kindly note that it is a requirement that you make use of [createasyncthunk](https://redux-toolkit.js.org/api/createAsyncThunk) when making your API calls according to the project requirements. I figured out that it was hard for you to implement this required change, I will give you an example trying to explain the moving parts as best and as briefly as I can.

```
import { createAsyncThunk } from '@reduxjs/toolkit';

export const fetchBooksList = createAsyncThunk(FETCH_BOOKS, async () => {
  const response = await fetch(API);
  const data = await response.json();

  const newArray = Object.keys(data).map((id) => {
    const obj = data[id][0];
    obj.id = id;
    return obj;
  });

  console.log(newArray);

  return newArray;
});
```

**Explanation**

1. You are first importing the createAsyncThunk into your file to use it. Make sure that you install it first.
2. You are then going to create the getBooks method using the createAsyncThunk. Here the GET_BOOK is simply the action type, nothing fancy
3. Then, because the data returned is of object data structure, I want to convert it into an array and return that array. This step is completely not a must and you could skip it and just return data.

I hope this very brief explanation helps you. Removing and adding a book with follow the same pattern as well.